core/staking_verifier.go

```diff
@@ -396,6 +397,104 @@ func VerifyAndUndelegateFromMsg(
 	return nil, errNoDelegationToUndelegate
 }
 
+// VerifyAndMigrateFromMsg verifies and transfers all delegations of
+// msg.From to msg.To. Returns all modified validator wrappers and delegate msgs
+// for metadata
+// Note that this function never updates the stateDB, it only reads from stateDB.
+func VerifyAndMigrateFromMsg(
+	stateDB vm.StateDB,
+	msg *staking.MigrationMsg,
+	fromDelegations []staking.DelegationIndex,
+) ([]*staking.ValidatorWrapper,
+	[]interface{},
+	error) {
+	if len(fromDelegations) == 0 {
+		// we are already done
+		return nil, nil, nil
+	}
+	if bytes.Equal(msg.From.Bytes(), msg.To.Bytes()) {
+		// we are already done
+		return nil, nil, nil
+	}
+	modifiedWrappers := make([]*staking.ValidatorWrapper, 0)
+	stakeMsgs := make([]interface{}, 0)
+	// iterate over all delegationIndexes by `From`
+	for i := range fromDelegations {
+		delegationIndex := &fromDelegations[i]
+		// find the wrapper for each delegationIndex
+		// request a copy, and since delegations will be changed, copy them too
+		wrapper, err := stateDB.ValidatorWrapper(delegationIndex.ValidatorAddress, false, true)
+		if err != nil {
+			return nil, nil, err
+		}
+		if uint64(len(wrapper.Delegations)) <= delegationIndex.Index {
+			utils.Logger().Warn().
+				Str("validator", delegationIndex.ValidatorAddress.String()).
+				Uint64("delegation index", delegationIndex.Index).
+				Int("delegations length", len(wrapper.Delegations)).
+				Msg("Delegation index out of bound")
+			return nil, nil, errors.New("Delegation index out of bound")
+		}
+		// and then find matching delegation to remove from wrapper
+		foundDelegation := &wrapper.Delegations[delegationIndex.Index] // note: pointer
+		if !bytes.Equal(foundDelegation.DelegatorAddress.Bytes(), msg.From.Bytes()) {
+			return nil, nil, errors.New(fmt.Sprintf("Expected %s but got %s",
+				msg.From.Hex(),
+				foundDelegation.DelegatorAddress.Hex()))
+		}
+		// Skip delegations with zero amount and empty undelegation
+		if foundDelegation.Amount.Cmp(common.Big0) == 0 && len(foundDelegation.Undelegations) == 0 {
+			continue
+		}
+		delegationAmountToMigrate := big.NewInt(0).Add(foundDelegation.Amount, big.NewInt(0))
+		undelegationsToMigrate := foundDelegation.Undelegations
+		// when undelegating we don't remove, just set the amount to zero
+		// to be coherent, do the same thing here (effective on wrapper since pointer)
+		foundDelegation.Amount = big.NewInt(0)
+		foundDelegation.Undelegations = make([]staking.Undelegation, 0)
+		// find `To` and give it to them
+		totalAmount := big.NewInt(0)
+		found := false
+		for i := range wrapper.Delegations {
+			delegation := &wrapper.Delegations[i]
+			if bytes.Equal(delegation.DelegatorAddress.Bytes(), msg.To.Bytes()) {
+				found = true
+				// add to existing delegation
+				totalAmount = delegation.Amount.Add(delegation.Amount, delegationAmountToMigrate)
+				// and the undelegation
+				delegation.Undelegations = append(delegation.Undelegations,
```

Comment:
Please follow the logic similar in VerifyAndUndelegateFromMsg. The Undelegations are based on epoch, so for undelegations of the same epoch, they should be merged.

Reply:
Done.